MAINT: Improve C level error messages for np.ones, zeros etc.

Kai-Striega opened the pull request:
This PR improves the error messages for TypeErrors in np.ones, np.ones_like, np.zeros, np.empty and np.empty_like by changing the error message of the corresponding C level functions array_zeros,  array_empty and array_empty_like.

```python
>>> import numpy as np
>>> np.zeros(1, 1)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: data type not understood, did you mean to use a tuple for size?
```
The issue comes from the dtype argument (in this case an integer value) not corresponding to a valid datatype. In which case the C function PyArray_DescrConverter cannot will raise register a TypeError. 

This issue was original raised on [Twitter](https://twitter.com/karpathy/status/1099793055853375489). Multiple PRs have been submitted [#13048, #13040], these have not been accepted as their focus  was on the matrixlib and that they worked at the Python level.

Kai-Striega committed:
ENH: Improve C level error messages for TypeErrors

This PR improves the error messages for TypeErrors in np.ones,
np.ones_like, np.zeros, np.empty and np.empty_like by changing the error
message of the corresponding C level functions array_zeros,  array_empty
and array_empty_like.

>>> import numpy as np
>>> np.zeros(1, 1)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: data type not understood, did you mean to use a tuple for size?

The issue comes from the dtype argument (in this case an integer value)
not corresponding to a valid datatype. In which case the C function
PyArray_DescrConverter cannot will raise register a TypeError.

This issue was original raised on Twitter [1]. Multiple PRs
have been submitted [2, 3], these have not been accepted as their focus
was on the matrixlib and that they worked at the Python level.

[1] https://twitter.com/karpathy/status/1099793055853375489
[2] https://github.com/numpy/numpy/pull/13048
[3] https://github.com/numpy/numpy/pull/13040

Kai-Striega commented on numpy/core/src/multiarray/multiarraymodule.c:
```diff
@@ -1839,6 +1839,11 @@ array_empty(PyObject *NPY_UNUSED(ignored), PyObject *args, PyObject *kwds)
+        PyErr_SetString(PyExc_TypeError,
+                "data type not understood, "
+                "did you mean to use a tuple for size?");
+    }
```
This attempts to match the TypeError set by PyArray_DescConverter where the passed value is not a valid type.

It cannot be moved any lower as the converter is used in many other places and the more specific error message may not be applicable to those cases

Kai-Striega commented on numpy/core/tests/test_api.py:
```diff
@@ -524,3 +524,18 @@ def test_broadcast_arrays():
+        " did you mean to use a tuple for size?")
+    with pytest.raises(TypeError) as e:
+        fn(1, 1)
+    assert_equal(str(e.value), message)
```
I'm not sure if here is the best location for the tests. Let me know if you would like me to move them.

Kai-Striega commented:
@mattip or @seberg I'm still new to working with the C code, would you be able to clarify the use of `.c` and `.c.src` files. I understand that they are used as Template files, but, should I be making the changes to ``numpy/core/src/multiarray/multiarraymodule.c`` or the corresponding ``.c.src`` file, even if I am not using any templates for the specific change?

Kai-Striega committed:
FIX: Explicitly clear PyErr when renaming error

eric-wieser reviewed (commented):
I'm not a massive fan of this, because:
* It throws away whatever detail the original `TypeError` had. Chaining exceptions would address this.
* I've not seen any precedent for _"did you mean to call this instead"_-style python exceptions.

seberg commented:
I agree with the first sentiment that this seems not specific enough. Also I am not actually sure how useful the current wording really is. I think I am OK with "did you mean to call this instead" type exceptions as such. It is true that I have never seen them in python either, but maybe it is a good idea to go towards longer error messages that give tips.

I feel this was a bit hyped, which is why if we get this I would like a solution that does not feel a bit slapped on. E.g. the type error here could also come from an array element being coerced! Frankly, it seems *more* likely to me. So error replacement should probably only occur if it is clear that the `TypeError` comes from parsing the `dtype` argument?

eric-wieser commented:
Perhaps a custom `PyArg_Parse...` "converter" argument should be used to parse the dtype?

seberg commented:
Maybe, or just catch the super common cases, by checking if the dtype python object was a python int/float...

Kai-Striega commented:
>  It throws away whatever detail the original TypeError had. Chaining exceptions would address this.

I'm not sure of the best way to daisy chain C level exceptions. Is there already an implementation somewhere in NumPy? If not, CPython's implementation [__PyErr_ChainExceptions(exc, val, tb)](https://hg.python.org/cpython/rev/9af21752ea2a) should be a good starting point. 

> I agree with the first sentiment that this seems not specific enough. Also I am not actually sure how useful the current wording really is.

I'm happy for suggestions on how to improve it =). The original message came from the previous PR and was acceptable by the discussion.

> I feel this was a bit hyped...

I agree that it *was* hyped. While, I don't think it is a major change technically, it occurs often enough for end-users to notice / become hyped. So a "fix" would help a lot of people a little bit.

> ...which is why if we get this I would like a solution that does not feel a bit slapped on.

Again, happy for any suggestions. I followed the suggestions from #13048, to look at ``array_{python name}`` in ``arraymodule.c``.

> Maybe, or just catch the super common cases, by checking if the dtype python object was a python int/float...

That seems like the simplest way to get it done. Would that be "proper" enough to be considered a long term solution?

Kai-Striega commented:
@seberg @eric-wieser a friendly bump

In summary, the main arguments against this are:

* Loses previous error message (does not chain exceptions)
* Custom error message is too ambiguous (or not wanted at all?)
* Error is too general (catch specific cases or add custom converted function)

I am happy to add all three but I would like to reach a consensus on how it should be done before I start making changes. Thoughts?

seberg commented:
> * Loses previous error message (does not chain exceptions)

This is easy, you can just chain the exception, we have a macro for that inside numpy which is identical to the newer macro in python.

> * Custom error message is too ambiguous (or not wanted at all?)

I am not against a custom error message. I am not even against a "Did you mean" style error message. But I feel if we do this, it should be useful even for people starting to learn numpy, so it has to dumb down a bit to be useful. Maybe with an example:
```
Did you mean to write `np.zeros((1, 2))` instead of `np.zeros(1, 2)` which translates to `np.zeros(1, dtype=2)`.
```
Just suggest something that seems too simple and it is probably god?

> * Error is too general (catch specific cases or add custom converted function)

I feel if what we really are scared of is the above, then we can just check if the type of the dtype input is an integer (or maybe number for `np.array` at least, but that is already much less likely). The other advantage of that is that exception chaining may not be necessary, because the error message is already super specific and can include that integers are not valid dtypes. (at least from a simplifying perspective I am not sure chained exceptions are all that nice).

> I am happy to add all three but I would like to reach a consensus on how it should be done before I start making changes. Thoughts?

I think: Just make a suggestion that errs on the safe side, and at least I would be happy.

eric-wieser commented:
One way to handle detecting the bad arguments might be to replace:
```
    if (!PyArg_ParseTupleAndKeywords(args, kwds, "O&|O&O&:zeros", kwlist,
                PyArray_IntpConverter, &shape,
                PyArray_DescrConverter, &typecode,
                PyArray_OrderConverter, &order)) {
        goto fail;
    }
```
with
```
    if (!PyArg_ParseTupleAndKeywords(args, kwds, "O&|O&O&:zeros", kwlist,
                PyArray_IntpConverter, &shape,
                PyArray_DescrConverterDetectingMisplacedShapeArgument, &typecode,
                PyArray_OrderConverter, &order)) {
        // special error type for detecting this unambiguously
        if (PyErr_Matches(PyArray_MisplacedShapeArgument)) {
            // replace the error with one containing all the arguments like @seberg suggests - no need to chain
        }
        goto fail;
    }
```

Kai-Striega commented:
@eric-wieser would it be enough for `PyArray_DescrConverterDetectingMisplacedShapeArgument` to wrap `PyArray_DescrConverter`?

What I mean `PyArray_DescrConverterDetectingMisplacedShapeArgument` is for  accept the arguments check if they are integral. If they are set the `PyArray_MisplacedShapeArgumentError` otherwise pass it on to ``PyArray_DescrConverter``

eric-wieser commented:
That sounds good to me

Kai-Striega commented on numpy/core/tests/test_api.py:
```diff
@@ -524,3 +524,18 @@ def test_broadcast_arrays():
     assert_equal(result[1], np.array([(1, 2, 3), (4, 5, 6), (7, 8, 9)], dtype='u4,u4,u4'))
+
+@pytest.mark.parametrize('fn', [np.ones, np.ones_like, np.zeros, np.empty, np.empty_like])
+def test_homogenous_array_creation_returns_descriptive_errror(fn):
```
TODO: Fix spelling error

Kai-Striega committed:
MAINT: Add custom DescrConv to check for Int types

This commit adds the DescrConverterDetectIntegerArgument function. It
checks if the passed dtype argument is of integer type. If it is a
BadArgument (i.e. a TypeError) is raised. If not the argument is passed
through to the standard DescrConverter.

The integer check is required to check if an invalid size was passed to
the array creation routines array_empty, array_zeros and
array_empty_like.

The size should either be an int (for single dimensions) or a tuple
(for multiple dimensions). For example a 2x2 shape should be entered as
np.ones((1, 1)), however it is often entered incorrectly as
np.ones(1, 1), in which case the second 1 is read as the dtype argument.

Kai-Striega committed:
MAINT: Add custom error for misplaced axis args

This commits adds a PyArray_MisplacedShapeArgumentError. The custom
error allows calling code to distinguish this case from the more general
TypeError currently raised.

Kai-Striega committed:
REVERT: Remove shape check for *_like functions

This commit removes the MisplacedShapeArgumentError for functions
ones_like, zeros_like, and empty_like. These require an array as input,
instead of an explicit shape argument. The MisplacedShapeArgumentError
is therefore not relevant.

Kai-Striega commented:
@seberg sorry about the delay. I was trying to pass the debug test only to find it was related to #14291.

For the error message what do you think of:

``np.{ones,zeros,empty} could not understand the dtype argument. If you want an array of shape (1, 1) this can occur if the shape is entered as np.{ones,zeros,empty}(1, 1) which is the same as np.{ones,zeros,empty}(1, dtype=1). Did you mean to use np.{ones,zeros,empty}((1, 1)) instead of np.{ones,zeros,empty}(1, 1)?``

seberg commented:
Seems pretty nice to me. I think I would like if the first sentence is identical to the original error (maybe plus `to/in np.{ones,...}`), which could be adjusted a bit if you like?

eric-wieser commented:
Can we include the actual values passed in the message?

seberg commented:
It appears that this PR was incorrectly closed by github while renaming the NumPy main (previously master) branch. I suspect because the original repository was also deleted.  That seems like a github issue, but I am not sure there is a solution to it aside from opening a new PR.

Sorry about that issue, I did not review the PR. However, even if the local branches were deleted, it seems the work is still available (e.g. using `https://github.com/numpy/numpy/pull/<pull-request-number>` with the command line interface), so we can assist creating a new PR!
